validator/resource_value_pattern.rego

```diff
@@ -0,0 +1,151 @@
+#
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPResourceValuePatternConstraintV1
+
+import data.validator.gcp.lib as lib
+
+###########################
+# Matches any field, including nested fields in asset data against
+# a regular expression pattern for validity checking.
+# fields can be optional such that missing fields in data
+# will not trigger a violation; or by default required
+# such that missing fields will trigger a violation.
+###########################
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	params := lib.get_constraint_params(constraint)
+	asset := input.asset
+
+	# asset type can be a scalar string or an array, only continue if the
+	# resource's asset type is one that should be looked at
+	# If asset types don't match exactly they will just be skipped
+	# potentially creating false positives.
+	# Ensure your assets types are 100% correct with negative testing.
+scalar_in(	asset.asset_type, params.asset_type)
```

Comment:
since params.asset_type is an array, do we really need the `scalar_in` function? would just `asset.asset_type == params.asset_type[_]` be sufficient?

Reply:
Addressed with latest commits.